pages/blog/react-as-generator-engine.md

```diff
@@ -0,0 +1,217 @@
+
+### Creating reusable parts
+
+It may sound obvious, but when writing any code, even a template, a programmer wants to create reusable parts that separate repeating logic.
+
+In Nunjucks, you can reuse parts of the template using **macros**, in React, using **components**. Imagine that you are writing a template that produces Markdown content. You need to create a reusable macro/component that renders a list from an array of strings.
+
+Using Nunjucks you can write the code below:
+
+```njk
+{% macro list(data, type = "-") %}
+{% for item in data %}
+{{type}} {{item}}
+{% endfor %}
+{% endmacro %}
+
+{% from "partials/list.njk" import list %}
+{{ list(["one", "two", "three"]) }}
+```
+
+Using React you can write the code below:
+
+```js
+function List({ list = [], type = "-" }) {
+  return list.map(item => `${type} ${item}\n`);
+}
+
+// use `List` component in another component
+export function SimpleList() {
+  return (
+    <List list={["one", "two", "three"]} />
+  );
+}
+```
+
+Looking at both examples we see that in Nujucks we operate on string literals, it means that when passing data to a macro, you always need to know what type of data the macro takes. In React we operate on JS objects/variables. By this, your IDE should always inform you what value, of what type, you must pass to component. Additionally, you must place Nunjucks's macro inside the **partials** folder of the template. Using React, you can put your components wherever you want.
+
+### Using third party packages
+
+Using helper functions from third party packages, in Nunjucks you must apply them as [filters](https://github.com/asyncapi/generator/blob/master/docs/authoring.md#filters). For example, you want to use one function from [`Underscore.string`](https://github.com/esamattis/underscore.string) library like `cleanDiacritics`, which replaces diacritic characters with closest ASCII equivalents. To do this, you must create a function inside `filters` folder to convert the function to Nunjucks's filter:
+
+```js
+// filters/cleanDiacritics.js
+const cleanDiacritics = require('underscore.string/cleanDiacritics');
+const filter = module.exports;
+filter.cleanDiacritics = cleanDiacritics;
+```
+
+And then you can use this function inside your template/macro:
+
+```njk
+{{ Urbańczyk | cleanDiacritics }} # will be Urbanczyk
+```
+
+The main problem in this solution is that it creates unnecessary boilerplate - you must create function in the separate file. Another problem is that you operate on name of this helper function which means you must always remember which filters you have included in your template.
```

Comment:
```suggestion
The main problem with this solution is that it creates an unnecessary boilerplate - you must create a function in a separate file and test it. Another problem is that you operate on the name of this helper function which means you must always remember what filters you have included in your template.
```

Reply:
Thanks, I will avoid this part `and test it`, because we don't mention about testing in react part. Important is that you must create function in seperate folder and only this has some value in this section.